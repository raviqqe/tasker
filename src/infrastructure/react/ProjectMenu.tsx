import { defaultImport } from "default-import";
import { useState, useRef, useEffect } from "react";
import { styled } from "@linaria/react";
import { type IProject } from "../../domain/project.js";
import {
  CreateProject,
  type IProps as ICreateProjectProps,
} from "./CreateProject.js";
import { Loader } from "./Loader.js";
import { Project, type IProps as IProjectProps } from "./Project.js";
import { ToggleProjects } from "./ToggleProjects.js";
import { grey, white, lightGrey } from "./style/colors.js";
import { boxShadow } from "./style.js";



const Container = styled.div`
  background-color: ${lightGrey};
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  height: 100vh;
`;

const ScrollContainer = styled.div`
  overflow: auto;
`;

const ProjectsContainer = styled.div`
  ${boxShadow}
  display: flex;
  flex-direction: column;
  margin: 1.5rem;
  max-width: 50ex;
  background-color: ${white};
  border-radius: 0.5rem;
  padding: 1rem 1.5rem;

  > :not(:first-child) {
    margin-top: 0.75rem;
  }
`;

const Message = styled.div`
  color: ${grey};
`;

const LowerButtonsContainer = styled.div`
  position: fixed;
  right: 0.5rem;
  bottom: 0.5rem;

  > * {
    margin-top: 0.5rem;
  }
`;

const StyledCreateProject = styled(CreateProject)<{
  projectsArchived: boolean;
}>`
  visibility: ${({ projectsArchived }) =>
    projectsArchived ? "hidden" : "visible"};
`;

export interface IProps
  extends ICreateProjectProps,
    Required<Omit<IProjectProps, "currentProject" | "project">> {
  archivedProjects: IProject[] | null;
  currentProject: IProject | null;
  hideProjects: () => void;
  projects: IProject[] | null;
}

export const ProjectMenu = ({
  archivedProjects,
  archiveProject,
  createProject,
  currentProject,
  deleteProject,
  hideProjects,
  projects,
  switchCurrentProject,
  unarchiveProject,
  updateProject,
}: IProps): JSX.Element => {
  const [projectsArchived, setProjectsArchived] = useState(false);
  const ref = useRef<HTMLDivElement>(null);

  useEffect(() => {
    if (ref.current) {
      ref.current.scrollIntoView({ block: "center" });
    }
  });

  return currentProject && archivedProjects && projects ? (
    <Container>
      <ScrollContainer>
        {projectsArchived ? (
          <ProjectsContainer>
            {archivedProjects.length === 0 ? (
              <Message>No archived projects</Message>
            ) : (
              archivedProjects.map((project) => (
                <Project
                  deleteProject={deleteProject}
                  key={project.id}
                  project={project}
                  unarchiveProject={async (project) => {
                    await unarchiveProject(project);
                    hideProjects();
                  }}
                />
              ))
            )}
          </ProjectsContainer>
        ) : (
          <ProjectsContainer>
            {projects.map((project) => (
              <Project
                archiveProject={archiveProject}
                currentProject={currentProject}
                key={project.id}
                project={project}
                ref={project.id === currentProject.id ? ref : null}
                switchCurrentProject={async (project) => {
                  hideProjects();
                  await switchCurrentProject(project);
                }}
                updateProject={updateProject}
              />
            ))}
          </ProjectsContainer>
        )}
      </ScrollContainer>
      <LowerButtonsContainer>
        <ToggleProjects
          projectsArchived={projectsArchived}
          setProjectsArchived={setProjectsArchived}
        />
        <StyledCreateProject
          createProject={async (name: string): Promise<void> => {
            await createProject(name);
            hideProjects();
          }}
          projectsArchived={projectsArchived}
        />
      </LowerButtonsContainer>
    </Container>
  ) : (
    <Container>
      <Loader />
    </Container>
  );
};
