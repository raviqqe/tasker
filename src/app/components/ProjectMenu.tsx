import { css } from "@linaria/core";
import { styled } from "@linaria/react";
import { useStore } from "@nanostores/react";
import { useEffect, useRef, useState } from "react";
import { MdArchive, MdDelete, MdEdit, MdUnarchive } from "react-icons/md";
import { projectArchiver } from "../../main/project-archiver.js";
import { projectCreator } from "../../main/project-creator.js";
import { projectDeleter } from "../../main/project-deleter.js";
import { projectPresenter } from "../../main/project-presenter.js";
import { projectUnarchiver } from "../../main/project-unarchiver.js";
import { projectUpdater } from "../../main/project-updater.js";
import { CreateProject } from "./CreateProject.js";
import { IconButton } from "./IconButton.js";
import { Loader } from "./Loader.js";
import { Project } from "./Project.js";
import { ToggleProjects } from "./ToggleProjects.js";
import { grey, lightGrey, white } from "./style.js";
import { boxShadow } from "./style.js";

const Container = styled.div`
  background-color: ${lightGrey};
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  height: 100vh;
`;

const ScrollContainer = styled.div`
  overflow: auto;
`;

const ProjectsContainer = styled.div`
  ${boxShadow}
  display: flex;
  flex-direction: column;
  margin: 1.5rem;
  max-width: 50ex;
  background-color: ${white};
  border-radius: 0.5rem;
  padding: 1rem 1.5rem;

  > :not(:first-child) {
    margin-top: 0.75rem;
  }
`;

const Message = styled.div`
  color: ${grey};
`;

const LowerButtonsContainer = styled.div`
  position: fixed;
  right: 0.5rem;
  bottom: 0.5rem;

  > * {
    margin-top: 0.5rem;
  }
`;

interface Props {
  onHideProjects: () => void;
}

export const ProjectMenu = ({ onHideProjects }: Props): JSX.Element => {
  const [projectsArchived, setProjectsArchived] = useState(false);
  const ref = useRef<HTMLDivElement>(null);
  const currentProject = useStore(projectPresenter.currentProject);
  const archivedProjects = useStore(projectPresenter.archivedProjects);
  const projects = useStore(projectPresenter.projects);

  useEffect(() => {
    if (ref.current) {
      ref.current.scrollIntoView({ block: "center" });
    }
  });

  return currentProject && archivedProjects && projects ? (
    <Container>
      <ScrollContainer>
        {projectsArchived ? (
          <ProjectsContainer>
            {archivedProjects.length === 0 ? (
              <Message>No archived projects</Message>
            ) : (
              archivedProjects.map((project) => (
                <Project
                  buttons={
                    <>
                      <IconButton
                        aria-label="Unarchive Project"
                        onClick={async () => {
                          await projectUnarchiver.unarchive(project);
                          onHideProjects();
                        }}
                      >
                        <MdUnarchive />
                      </IconButton>
                      <IconButton
                        aria-label="Delete Project"
                        onClick={() => projectDeleter.delete(project)}
                      >
                        <MdDelete />
                      </IconButton>
                    </>
                  }
                  currentProject={currentProject}
                  key={project.id}
                  project={project}
                />
              ))
            )}
          </ProjectsContainer>
        ) : (
          <ProjectsContainer>
            {projects.map((project) => (
              <Project
                buttons={
                  <>
                    <IconButton
                      aria-label="Update Project"
                      onClick={async () => {
                        const name = window.prompt(
                          "New project name?",
                          project.name,
                        );

                        if (!name) {
                          return;
                        }

                        await projectUpdater.update({ ...project, name });
                      }}
                    >
                      <MdEdit />
                    </IconButton>
                    <IconButton
                      aria-label="Archive project"
                      onClick={async () => {
                        await projectArchiver.archive(
                          project,
                          currentProject.id,
                        );
                      }}
                    >
                      <MdArchive />
                    </IconButton>
                  </>
                }
                currentProject={currentProject}
                key={project.id}
                onSwitchProject={onHideProjects}
                project={project}
                ref={project.id === currentProject.id ? ref : null}
              />
            ))}
          </ProjectsContainer>
        )}
      </ScrollContainer>
      <LowerButtonsContainer>
        <ToggleProjects
          projectsArchived={projectsArchived}
          setProjectsArchived={setProjectsArchived}
        />
        <CreateProject
          className={
            projectsArchived
              ? css`
                  visibility: hidden;
                `
              : undefined
          }
          onCreate={async (name) => {
            await projectCreator.create(name);
            onHideProjects();
          }}
        />
      </LowerButtonsContainer>
    </Container>
  ) : (
    <Container>
      <Loader />
    </Container>
  );
};
