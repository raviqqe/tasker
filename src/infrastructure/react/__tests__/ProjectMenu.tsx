import React from "react";
import { create } from "react-test-renderer";
import { ProjectMenu } from "../ProjectMenu";

it("renders", () => {
  expect(
    create(
      <ProjectMenu
        archivedProjects={[]}
        archiveProject={async () => {}}
        createProject={async () => {}}
        currentProject={{ archived: false, id: "", name: "" }}
        deleteProject={async () => {}}
        hideProjects={() => {}}
        projects={[]}
        switchCurrentProject={async () => {}}
        unarchiveProject={async () => {}}
        updateProject={async () => {}}
      />
    ).toJSON()
  ).toMatchSnapshot();
});
