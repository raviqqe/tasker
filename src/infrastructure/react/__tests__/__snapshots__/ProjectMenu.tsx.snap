// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<div
  className="sc-AykKI fqZTbk"
>
  <div
    className="sc-AykKJ jNGRFD"
  >
    <div
      className="sc-AykKK eTZoLT"
    />
  </div>
  <div
    className="sc-fzXfLP eYXeVw"
  >
    <button
      aria-label="Toggle Projects"
      className="sc-AykKC sc-AykKD ckEQXK"
      onClick={[Function]}
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M20.55 5.22l-1.39-1.68C18.88 3.21 18.47 3 18 3H6c-.47 0-.88.21-1.15.55L3.46 5.22C3.17 5.57 3 6.01 3 6.5V19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V6.5c0-.49-.17-.93-.45-1.28zM12 9.5l5.5 5.5H14v2h-4v-2H6.5L12 9.5zM5.12 5l.82-1h12l.93 1H5.12z"
        />
      </svg>
    </button>
    <button
      aria-label="Create Project"
      className="sc-AykKC sc-AykKD eoJpuV sc-fzXfLQ hHINjI"
      onClick={[Function]}
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
        />
      </svg>
    </button>
  </div>
</div>
`;
